SWServerToContextConnection::terminateWhenPossible should ask UIProcess to disable service workers on the corresponding WebProcess

youennf opened the pull request:
#### 7b303b1c7b274341cc5757fd1ede764f100559d3
<pre>
SWServerToContextConnection::terminateWhenPossible should ask UIProcess to disable service workers on the corresponding WebProcess
<a href="https://rdar.apple.com/139968744">rdar://139968744</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=283173">https://bugs.webkit.org/show_bug.cgi?id=283173</a>

Reviewed by Chris Dumez.

UIProcess may sometimes ask network process to stop using a web process for service workers.
This ends up calling SWServerToContextConnection::terminateWhenPossible.
When feasible, network process will then ask the web process to close its service workers and will think it no longer has a web process for running service workers.
But network process forgets to tell UIProcess about it so UIProcess does not call disableRemoteWorkers on the corresponding WebProcessProxy.

When network process needs a process for running service workers, it will ask UIProcess for a web process and UIProcess may reuse that process.
Normally it should not reuse it since UIProcess asked network process to stop using this web process for service workers as it was dirtied.
But this may be happening and UIProcess may terminate service workers for other reasons as well.

For that reason, instead of letting network process stops service workers directly, we make it go through UIProcess using the usual connectionIsNoLongerNeeded code path.
It is then UIProcess that will tell the web process to stop the service workers.

* Source/WebCore/workers/service/server/SWServer.cpp:
(WebCore::SWServer::terminateContextConnectionWhenPossible):
* Source/WebCore/workers/service/server/SWServerToContextConnection.cpp:
(WebCore::SWServerToContextConnection::terminateWhenPossible):
* Source/WebCore/workers/service/server/SWServerToContextConnection.h:
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerToContextConnection.cpp:
(WebKit::WebSWServerToContextConnection::close): Deleted.
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerToContextConnection.h:

Canonical link: <a href="https://commits.webkit.org/286849@main">https://commits.webkit.org/286849@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/69fe9b5b036b687ab8421151dacab666ca0a602a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/77174 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/56209 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/30089 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/81735 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/28456 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/79291 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/65357 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/4505 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/60461 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18522 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/80241 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/50430 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/66238 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/40764 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47832 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/23736 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/26779 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68947 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/24071 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/83161 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/4554 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/3066 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/68750 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/4710 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/66211 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/68006 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16999 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11984 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/10074 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/4501 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/7316 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/4520 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/7955 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/6279 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/d5f5ecfd8fac9d6d95058c4fdc9376d7c604ccb2. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/36697

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/69fe9b5b036b687ab8421151dacab666ca0a602a. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/36697

youennf committed:
SWServerToContextConnection::terminateWhenPossible should ask UIProcess to disable service workers on the corresponding WebProcess
rdar://139968744
https://bugs.webkit.org/show_bug.cgi?id=283173

Reviewed by Chris Dumez.

UIProcess may sometimes ask network process to stop using a web process for service workers.
This ends up calling SWServerToContextConnection::terminateWhenPossible.
When feasible, network process will then ask the web process to close its service workers and will think it no longer has a web process for running service workers.
But network process forgets to tell UIProcess about it so UIProcess does not call disableRemoteWorkers on the corresponding WebProcessProxy.

When network process needs a process for running service workers, it will ask UIProcess for a web process and UIProcess may reuse that process.
Normally it should not reuse it since UIProcess asked network process to stop using this web process for service workers as it was dirtied.
But this may be happening and UIProcess may terminate service workers for other reasons as well.

For that reason, instead of letting network process stops service workers directly, we make it go through UIProcess using the usual connectionIsNoLongerNeeded code path.
It is then UIProcess that will tell the web process to stop the service workers.

* Source/WebCore/workers/service/server/SWServer.cpp:
(WebCore::SWServer::terminateContextConnectionWhenPossible):
* Source/WebCore/workers/service/server/SWServerToContextConnection.cpp:
(WebCore::SWServerToContextConnection::terminateWhenPossible):
* Source/WebCore/workers/service/server/SWServerToContextConnection.h:
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerToContextConnection.cpp:
(WebKit::WebSWServerToContextConnection::close): Deleted.
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerToContextConnection.h:

Canonical link: https://commits.webkit.org/286849@main

webkit-commit-queue commented:
Committed 286849@main (7b303b1c7b27): <https://commits.webkit.org/286849@main>

Reviewed commits have been landed. Closing PR #36697 and removing active labels.
